clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractHeartbeatRequestManager.java

```diff
@@ -375,81 +342,88 @@ private void onErrorResponse(final ConsumerGroupHeartbeatResponse response,
                 handleFatalFailure(error.exception(errorMessage));
                 break;
 
             case FENCED_MEMBER_EPOCH:
-                message = String.format("GroupHeartbeatRequest failed for member %s because epoch %s is fenced.",
-                        membershipManager.memberId(), membershipManager.memberEpoch());
+                message = String.format("%s failed for member %s because epoch %s is fenced.",
+                        heartbeatRequestName(), membershipManager().memberId(), membershipManager().memberEpoch());
                 logInfo(message, response, currentTimeMs);
-                membershipManager.transitionToFenced();
+                membershipManager().transitionToFenced();
                 // Skip backoff so that a next HB to rejoin is sent as soon as the fenced member releases its assignment
                 heartbeatRequestState.reset();
                 break;
 
             case UNKNOWN_MEMBER_ID:
-                message = String.format("GroupHeartbeatRequest failed because member %s is unknown.",
-                        membershipManager.memberId());
+                message = String.format("%s failed because member %s is unknown.",
+                        heartbeatRequestName(), membershipManager().memberId());
                 logInfo(message, response, currentTimeMs);
-                membershipManager.transitionToFenced();
+                membershipManager().transitionToFenced();
                 // Skip backoff so that a next HB to rejoin is sent as soon as the fenced member releases its assignment
                 heartbeatRequestState.reset();
                 break;
 
             default:
                 // If the manager receives an unknown error - there could be a bug in the code or a new error code
-                logger.error("GroupHeartbeatRequest failed due to unexpected error {}: {}", error, errorMessage);
-                handleFatalFailure(error.exception(errorMessage));
+                if (!handleSpecificError(response, currentTimeMs)) {
+                    logger.error("{} failed due to unexpected error {}: {}", heartbeatRequestName(), error, errorMessage);
+                    handleFatalFailure(error.exception(errorMessage));
+                }
                 break;
         }
     }
 
-    private void logInfo(final String message,
-                         final ConsumerGroupHeartbeatResponse response,
+    protected void logInfo(final String message,
+                         final HBR response,
                          final long currentTimeMs) {
         logger.info("{} in {}ms: {}",
-            message,
-            heartbeatRequestState.remainingBackoffMs(currentTimeMs),
-            response.data().errorMessage());
+                message,
+                heartbeatRequestState.remainingBackoffMs(currentTimeMs),
+                errorMessageForResponse(response));
```

Comment:
indentation with 8 instead of 4 continuation that we had unintentionally in the previous PR, still unintentional here?